### PR TITLE
feat(comp: affix): add affix component

### LINKED ISSUE #138: [comp: affix] add affix component
- [x] I have searched the [issues](https://github.com/IduxFE/components/issues) of this repository and believe that this is not a duplicate.

### What problem does this feature solve?

affix

### props
|     名称     |                             说明                             |    类型     | 默认值 |
| :----------: | :----------------------------------------------------------: | :---------: | :----: |
|  offset  |         距离窗口顶部的偏移量，到达时会触发fix<br />          |   number \| { top:number. bottom:number, left:number, right:number}   |   -    |
|  container   |                        用于定位的容器，会监听容器的滚动事件，用函数返回        | () => HTMLElement \| window | () => window |
| throttleTime|   节流时间（ms）设置，用于提高性能<br />开启后affix位置更新会产生延迟   |   number    | -|

### event
|      名称      |            说明             | 回调参数 |
| :------------: | :-------------------------: | :------: |
| status-changed | 固定/非固定状态切换时的事件 | boolean  |

### slot
|      名称      |            说明             |
| :------------: | :-------------------------: |
| default | 默认插槽 |

### PR DESCRIPTION
**Pull request by jackson-yyy**
fix #138 

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows [our guidelines](https://github.com/IduxFE/idux/blob/main/docs/contributing.zh.md#commit)
- [x] Tests for the changes have been added/updated or not needed
- [x] Docs and demo have been added/updated or not needed

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Component style update
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Application (the showcase website) / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

don't have a affix component

## What is the new behavior?
have a affix component

### CODE REVIEW

File: packages/components/affix/src/Affix.vue [hunk 1]
```diff
@@ -0,0 +1,133 @@
+<template>
+  <div ref="wrapperRef" :style="wrapperStyle" class="ix-affix-wrapper">
+    <div ref="affixRef" class="ix-affix" :style="affixStyle">
+      <slot></slot>
+    </div>
+  </div>
```
Comment (danranVm) [inline packages/components/affix/src/Affix.vue:6]:
```suggestion
  <div ref="affixRef" :style="affixStyle" class="ix-affix">
    <div ref="contentRef" class="ix-affix-content" :style="contentStyle">
      <slot></slot>
    </div>
  </div>
```

最外层的 class 名和组件同名，整体组件都保持这个风格吧。

File: packages/components/affix/src/Affix.vue [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    const affixRef = ref<null | HTMLElement>(null)
+    const wrapperRef = ref<null | HTMLElement>(null)
+    const wrapperStyle = ref({})
+    const offsetOpt = computed(() => {
+      const offset = props.offset
+      if (typeof offset === 'number' || typeof offset === 'string') {
+        return { top: +offset }
+      }
+      return isEmpty(offset) ? { top: 0 } : offset
+    })
```
Comment (danranVm) [inline packages/components/affix/src/Affix.vue:40]:
```suggestion
    const offset = computed(() => {
      const _offset = props.offset
      if(isObject(_offset)){
        return _offset
      }
      return { top: toNumber(_offset) }
    })
```

File: packages/components/affix/src/Affix.vue [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    const isWindow = computed(() => !isHTMLElement(container.value))
```
Comment (danranVm) [inline packages/components/affix/src/Affix.vue:41]: 这个就不要了吧，在用的时候判断一下就好了 `container === window`

File: packages/components/affix/src/Affix.vue [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+const direction: Direction[] = ['top', 'bottom', 'left', 'right']
+
+const eventList = ['scroll', 'resize']
+
+export default defineComponent({
+  name: 'IxAffix',
+  props: {
+    offset: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.object]).def(0),
+    container: PropTypes.oneOfType([PropTypes.string, HTMLElement, Window]).def(() => window),
```
Comment (danranVm) [inline packages/components/affix/src/Affix.vue:24]:
```suggestion
    target: withUndefined(PropTypes.oneOfType([PropTypes.string, HTMLElement])),
```
跟 backtop 保持一致吧

File: packages/components/affix/src/Affix.vue [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    onUnmounted(removeListeners)
+
+    watch(() => props.offset, measure)
+    watch(
+      () => props.container,
+      () => {
+        removeListeners()
+        initContainer()
+        measure()
+      },
```
Comment (danranVm) [inline packages/components/affix/src/Affix.vue:61]:
```suggestion
      (currTarget, prevTarger) => {
        removeListeners(prevTarger)
        initContainer()
        measure()
      },
```

**Review by danranVm — CHANGES_REQUESTED** [review]
可以多参考一下 backtop 组件

**Review by Pupils-web — COMMENTED** [review]
已审

File: packages/components/affix/demo/Basic.vue [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  setup() {
+    const offset = ref(0)
+    return {
+      offset,
+    }
+  },
+})
+</script>
+<style lang="less" scoped>
+/deep/.ix-affix {
```
Comment (danranVm) [inline packages/components/affix/demo/Basic.vue:19]:
```suggestion
:deep(.ix-affix) {
```

File: packages/components/affix/src/utils.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
+import { isHTMLElement, isUndefined, isObject, toNumber } from '@idux/cdk/utils'
+import { on, off } from '@idux/cdk/utils'
+
+import { Direction, AffixStyle, OffsetOpt } from './types'
+
+const EVENT_LIST = ['scroll', 'resize']
+const DIRECTION: Direction[] = ['top', 'bottom', 'left', 'right']
```
Comment (danranVm) [inline packages/components/affix/src/utils.ts:7]:
```suggestion
const eventList = ['scroll', 'resize']
const directions: Direction[] = ['top', 'bottom', 'left', 'right']
```
只有跨模块使用的常量才需要全大写命名

File: packages/components/affix/src/utils.ts [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+type DirectionNumber = { [key in Direction]: number }
+
+export function formatOffset(offset: number | string | OffsetOpt): Partial<DirectionNumber> {
+  if (!isObject(offset)) {
+    return { top: toNumber(offset) }
+  }
+  const _offset: Partial<DirectionNumber> = {}
+  ;(Object.getOwnPropertyNames(offset) as Direction[]).forEach(dire => {
```
Comment (danranVm) [inline packages/components/affix/src/utils.ts:16]:
```suggestion
  Object.keys(offset).forEach(dire => {
```

File: packages/components/affix/__tests__/util.spec.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
+import { getTargetSize, getContainer, rafFn } from '../src/utils'
```
Comment (danranVm) [inline packages/components/affix/__tests__/util.spec.ts:1]: 组件内部的 utils 可以不需要单元测试的，按理说你在测试组件的时候，这些场景都会跑到。

File: packages/components/affix/src/Affix.vue [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+  removeObserveTarget,
+} from './utils'
+import { getTarget } from '@idux/components/core/utils/propsHandler'
+
+export default defineComponent({
+  name: 'IxAffix',
+  props: {
+    offset: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.object]).def(0),
+    target: withUndefined(PropTypes.oneOfType([PropTypes.string, HTMLElement])),
+    throttleTime: PropTypes.number,
```
Comment (danranVm) [inline packages/components/affix/src/Affix.vue:28]: 这玩意不是说移除了吗。还有文档也记得删一下。

File: packages/components/core/utils/propsHandler.ts [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+
+export function getTarget(target: string | HTMLElement | Window = window): HTMLElement | Window {
+  const targetDom = getTargetDom(target)
+  if (targetDom) {
+    return targetDom
+  } else {
+    Logger.warn(`target does not exist: ${target}, default value are already used: window.`)
+    return window
+  }
+}
```
Comment (danranVm) [inline packages/components/core/utils/propsHandler.ts:12]:
```suggestion
const getTarget = (target: HTMLElement | string | undefined) => {
  if (isHTMLElement(target)) {
    return target
  }
  if (isString(target)) {
    const container = document.querySelector<HTMLElement>(target)
    if (container) {
      return container
    } else {
      Logger.warn(`target does not exist: ${target}, default value are already used: window.`)
    }
  }
  return window
}
```

cdk 里的那个删掉吧，另外记得在 core/utils/index.ts 里 export 一下。

File: packages/components/back-top/src/BackTop.vue [hunk 12]
```diff
@@ -15,10 +15,10 @@ import type { BackTopProps } from './types'
 import { defineComponent, nextTick, onUnmounted, onMounted, ref } from 'vue'
 import throttle from 'lodash/throttle'
 import { IxIcon } from '@idux/components/icon'
-import { PropTypes, withUndefined, isString, isHTMLElement, on, off, rAF } from '@idux/cdk/utils'
+import { PropTypes, withUndefined, on, off, rAF } from '@idux/cdk/utils'
 import { easeInOutQuad } from '@idux/components/core/utils'
 import { useGlobalConfig } from '@idux/components/core/config'
-import { Logger } from '@idux/components/core/logger'
+import { getTarget } from '@idux/components/core/utils/propsHandler'
```
Comment (danranVm) [inline packages/components/back-top/src/BackTop.vue:21]:
```suggestion
import { getTarget } from '@idux/components/core/utils'
```

跟上面的 easeInOutQuad  合到一起吧

File: packages/components/affix/src/Affix.vue [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+import {
+  getTargetRect,
+  getTargetSize,
+  calcPosition,
+  formatOffset,
+  isSticky as judgeSticky,
+  observeTarget,
+  removeObserveTarget,
+} from './utils'
+import { getTarget } from '@idux/components/core/utils/propsHandler'
```
Comment (danranVm) [inline packages/components/affix/src/Affix.vue:21]:
```suggestion
import { getTarget } from '@idux/components/core/utils'
```

**Review by danranVm — CHANGES_REQUESTED** [review]

**Review by danranVm — APPROVED** [review]
LGTM.

**Pull request merged**